log/DESIGN.md

```diff
@@ -0,0 +1,380 @@
+- `go.opentelemetry.io/otel/log/noop`
+
+Rejected alternative:
+
+- [Reuse slog](#reuse-slog)
+
+### LoggerProvider
+
+The [`LoggerProvider` abstraction](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#loggerprovider)
+is defined as an interface:
+
+```go
+type LoggerProvider interface {
+	embedded.LoggerProvider
+	Logger(name string, options ...LoggerOption) Logger
+}
+```
+
+The specification may add new operations to `LoggerProvider`.
+The interface may have methods added without a package major version bump.
+The embedds `embedded.LoggerProvider` to help inform an API implementation
+author about this non-standard API evolution.
+This approach is already used in Trace API and Metrics API.
+
+#### LoggerProvider.Logger
+
+The `Logger` method implements the [`Get a Logger` operation](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#get-a-logger).
+
+Implementation requirements:
+
+- The [specification requires](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#concurrency-requirements)
+  the method to be safe to be called concurrently.
+
+- The method should use some default name if the passed name is empty
+  in order to meet the [specification's SDK requirement](https://opentelemetry.io/docs/specs/otel/logs/sdk/#logger-creation)
+  to return a working logger when an invalid name is passed
+  as well as to resemble the behavior of getting tracers and meters.
+
+`Logger` can be extended by adding new `LoggerOption` options
+and adding new exported fields to the `LoggerConfig` struct.
+This design is already used in Trace API for getting tracers
+and in Metrics API for getting meters.
+
+Rejected alternative:
+
+- [Passing struct as parameter to LoggerProvider.Logger](#passing-struct-as-parameter-to-loggerproviderlogger).
+
+### Logger
+
+The [`Logger` abstraction](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#logger)
+is defined as an interface:
+
+```go
+type Logger interface {
+	embedded.Logger
+	Emit(ctx context.Context, record Record)
+}
+```
+
+The specification may add new operations to `Logger`.
+The interface may have methods added without a package major version bump.
+The embedds `embedded.Logger` to help inform an API implementation
+author about this non-standard API evolution.
+This approach is already used in Trace API and Metrics API.
+
+### Logger.Emit
+
+The `Emit` method implements the [`Emit a LogRecord` operation](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#emit-a-logrecord).
+
+Calls to `Emit` are supposed to be on the hot path.
+Therefore, in order to reduce the number of heap allocations,
+the [`LogRecord` abstraction](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#emit-a-logrecord),
+is defined as a `struct`:
+
+```go
+type Record struct {
+	Timestamp         time.Time
+	ObservedTimestamp time.Time
+	Severity          Severity
+	SeverityText      string
+	Body              string
+	Attributes        []attribute.KeyValue
```

Comment:
One question I have is will our current `attribute.Keyvalue` work as the [Log version of attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-attributes).  Currently the Spec defines these as `Map<string, any>`, where [`any`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#type-any) can be, scaler (our int, float, string, bool), byte slice (string), an array of the above, or a Map<string, any> (we don't support).

Would that mean we would need to add support for that into the attribute package, and if so, what's the effect on other signals that don't allow that?  
If we don't, will we have to add a conversion from attributes to log attributes?

Reply:
Great finding that I have totally missed. I am not sure at the moment as https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/README.md#attribute does not say anyything about having maps in attributes. I have also found https://github.com/open-telemetry/opentelemetry-specification/issues/376 and https://github.com/open-telemetry/opentelemetry-specification/issues/622. Thoughts?

Reply:
Java (stable logs): [reuses a common attribute type](https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/logs/LogRecordBuilder.java) and [a map as value is not supported](https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/common/AttributeType.java)

.NET (stable logs): [models attributes](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Api/Logs/LogRecordAttributeList.cs) (everywhere) as [`KeyValuePair<string, object?>`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.keyvaluepair-2)

C++ (stable logs): [reuses a common attribute type](https://github.com/open-telemetry/opentelemetry-cpp/blob/main/api/include/opentelemetry/logs/log_record.h) and [a map as value is not supported](https://github.com/open-telemetry/opentelemetry-cpp/blob/main/api/include/opentelemetry/common/attribute_value.h)

Python (experimental logs): [reuses a common attribute type](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-sdk/src/opentelemetry/sdk/_logs/_internal/__init__.py) and [a map as value is not supported](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-api/src/opentelemetry/util/types.py)

I am leaning that it is a bug in the specification and that attributes values do not currently accept maps as values. If it would be the case then it should work for all signals.

Reply:
CC @tigrannajaryan

Reply:
I think (I am aware that I can be totally wrong) that in https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md:
- `Attributes` should be defined as (reference) https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/common
- `Resource` should be defined as (reference) https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md

Reply:
>Attributes should be simply defined as (reference) https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/common

That won't work. Common attributes don't allow nesting. Log attributes (and Body) allow it.

Maybe this is another opportunity to push for resolution on https://github.com/open-telemetry/opentelemetry-specification/issues/376 and make common attribute definition equal to `any` type definition in log data model.

Reply:
Maybe this is another opportunity to push for resolution on https://github.com/open-telemetry/opentelemetry-specification/issues/376

Reply:
> Maybe this is another opportunity to push for resolution on open-telemetry/opentelemetry-specification#376

I agree as currently none languages (expect .NET) does not allow nesting.

Reply:
@tigrannajaryan  As far as I checked NONE languages support nested log attributes. All languages reuses common attributes abstraction (Java, C++, Python) or transformation (.NET).

I think we have a big precedence that log attributes may use the same representation as other signals and that  addressing https://github.com/open-telemetry/opentelemetry-specification/issues/376 would solve the problem for all languages.

Reply:
I created https://github.com/open-telemetry/opentelemetry-specification/pull/3816

Reply:
I plan to create define dedicated type for log attributes. If https://github.com/open-telemetry/opentelemetry-specification/pull/3816 will be accepted then I will simply change it back to use the `attribute` package. I will open an issue for tracking.

Reply:
PR for the prototype https://github.com/pellared/opentelemetry-go/pull/7